### PR TITLE
SCB-224 retry sub-transaction on failure

### PR DESCRIPTION
**Pull request by eric-lee-ltk**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

This PR is based on #117

### CODE REVIEW

**Comment by WillemJiang** [comment]
@eric-lee-ltk  There are some build error in you commit, could you take a look at it?

File: omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/DefaultRecovery.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import javax.transaction.InvalidTransactionException;
+
+import org.apache.servicecomb.saga.omega.context.OmegaContext;
+import org.apache.servicecomb.saga.omega.transaction.annotations.Compensable;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultRecovery implements RecoveryPolicy {
```
[the remark this answers is not in the document]
Comment (eric-lee-ltk) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/DefaultRecovery.java]: Not exactly. As the compensation of backward recovery is handled asynchronously, this place is only used to execute bussiness logic once.

File: omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/RecoveryPolicyFactory.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package org.apache.servicecomb.saga.omega.transaction;
+
+public class RecoveryPolicyFactory {
+  private static final RecoveryPolicy defaultRecovery = new DefaultRecovery();
+
+  private static final RecoveryPolicy forwardRecovery = new ForwardRecovery();
+
+  static RecoveryPolicy getRecoveryPolicy(int retries) {
```
[the remark this answers is not in the document]
Comment (eric-lee-ltk) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/RecoveryPolicyFactory.java]: Which parameter is more proper?

[the remark this answers is not in the document]
Comment (WillemJiang) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/RecoveryPolicyFactory.java]: It depends on the RecoveryPolicy that you have,  maybe you need to add some comments for the DefaultRecovery.

File: omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          if (remains == 0) {
+            log.error(
+                "Retried sub tx failed maximum times, global tx id: {}, local tx id: {}, method: {}, retried times: {}",
+                context.globalTxId(), context.localTxId(), method.toString(), retries);
+            throw throwable;
+          }
+
+          log.warn("Retrying sub tx failed, global tx id: {}, local tx id: {}, method: {}, remains: {}",
+              context.globalTxId(), context.localTxId(), method.toString(), remains);
+          Thread.sleep(compensable.retryDelayInMilliseconds());
```
[the remark this answers is not in the document]
Comment (eric-lee-ltk) [reply omega/omega-transaction/src/main/java/org/apache/servicecomb/saga/omega/transaction/ForwardRecovery.java]: Scheduler requires to implement the Callable interface to retrieve the return value. However, we need to construct a new instance for every execution to pass the paramenters. The times of execution is also hard to control. I guess this also introduce performance penalty.

**Comment by WillemJiang** [comment]
Merged the patch into master branch.

**Pull request merged**